src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1395,7 +1395,24 @@ internal bool HasTopLevelNullabilityIdentityConversion(TypeSymbolWithAnnotations
             {
                 return true;
             }
-            return source.IsNullable == destination.IsNullable;
+
+            if (source.NullableAnnotation == NullableAnnotation.Unknown ||
+                destination.NullableAnnotation == NullableAnnotation.Unknown)
+            {
+                return true;
+            }
+
+            if (source.IsPossiblyNullableReferenceTypeTypeParameter() && !destination.IsPossiblyNullableReferenceTypeTypeParameter())
+            {
+                return destination.NullableAnnotation.IsAnyNullable();
+            }
+
+            if (destination.IsPossiblyNullableReferenceTypeTypeParameter() && !source.IsPossiblyNullableReferenceTypeTypeParameter())
+            {
+                return source.NullableAnnotation.IsAnyNullable();
+            }
+
+            return source.NullableAnnotation.IsAnyNullable() == destination.NullableAnnotation.IsAnyNullable();
```

Comment:
How would describe the purpose/semantics of `HasTopLevelNullabilityIdentityConversion` for an xml doc?

I can think of two ways:
1. returns true if you can assign from source to destination and from destination to source (in terms of top-level nullability)
2. returns true if the value-nullabilities of source and destination match, with the exception of oblivious (matches anything)

If we prefer the first description, then it is natural to express it in terms of `HasTopLevelNullabilityImplicitConversion` as I proposed earlier.
If we prefer the second description, then this method can be expressed simply by checking for oblivious and then comparing `ValueCanBeNull` for the source and the destination.

If neither of these descriptions is correct, or there is a better one, such a description will certainly help to come to an understanding of the purpose of this method, so that it can effectively be reviewed.

Reply:
> How would describe the purpose/semantics of HasTopLevelNullabilityIdentityConversion for an xml doc?

Adding a doc-comment in the next commit

---
In reply to: [231454173](https://github.com/dotnet/roslyn/pull/30913#discussion_r231454173) [](ancestors = 231454173)

Reply:
Do you mind sharing what it will say, so that we are on the same page for this code in the current PR? Thanks
My sense is that once we express the intent clearly, we likely will be able to simplify the code to reflect that intent. 

---
In reply to: [231634476](https://github.com/dotnet/roslyn/pull/30913#discussion_r231634476) [](ancestors = 231634476,231454173)

Reply:
> Do you mind sharing what it will say, so that we are on the same page for this code in the current PR?

```
        /// <summary>
        /// Returns true if:
        /// - Either type has no nullability information (oblivious).
        /// - Both types cannot have different nullability at the same time.
        /// </summary>
```

---
In reply to: [231670473](https://github.com/dotnet/roslyn/pull/30913#discussion_r231670473) [](ancestors = 231670473,231634476,231454173)

Reply:
What do you think of doing the following then?
```C#
            if (source.NullableAnnotation == NullableAnnotation.Unknown ||
                destination.NullableAnnotation == NullableAnnotation.Unknown)
            {
                return true;
            }
            return source.ValueCanBeNull() == destination.ValueCanBeNull();
```

Could you also share your proposed description for `HasTopLevelNullabilityImplicitConversion`?

---
In reply to: [231688354](https://github.com/dotnet/roslyn/pull/30913#discussion_r231688354) [](ancestors = 231688354,231670473,231634476,231454173)

Reply:
Following our discussion, I tried that alternative code for `HasTopLevelNullabilityIdentityConversion`.
It passed the tests, but due to my experiment on `HasTopLevelNullabilityImplicitConversion`, I now think this alternative is likely incorrect. More on that in next comment.

Reply:
> It passed the tests, but due to my experiment on HasTopLevelNullabilityImplicitConversion, I now think this alternative is likely incorrect. More on that in next comment.

Yes, the proposed alternative fails to distinguish T? from T. When T is replaced with a non-nullable type, we are dealing with types of different nullability. The ValueCanBeNull helper assumes the "worst" case and ignores the fact that T could actually be not nullable. These functions should consider both scenarios.   


---
In reply to: [231760271](https://github.com/dotnet/roslyn/pull/30913#discussion_r231760271) [](ancestors = 231760271)

Reply:
For the record, some additional details on this scenario:
`T` is constrained to `class?`.
During analysis we could end up with a `T?` (NullableAnnotation.Nullable, for example by assigning null to something), as well as a `T` (NullableAnnotation.Unannotated, and also “possibly nullable reference type type parameters”).
If `T` were `object`, it is invalid to assign the value from the `T?` to the `T`.
But methods like `GetValueNullableAnnotation` and `CanValueBeNull` treat both the same, ie. CanValueBeNull is true.

---
In reply to: [231771040](https://github.com/dotnet/roslyn/pull/30913#discussion_r231771040) [](ancestors = 231771040,231760271)